Added missing parts

Delphi-FPC-Lazarus opened the pull request:
Changes: (see checkin comments for details)
- fixed warning cause of comparison signed and unsinged integer
- ifdef inline will lead to an F2051 on complex nested package structures containing runtime and design packages
- Added properties for accessing context. Needed for calling SSL_get_*() functions like SSL_get_version()
- added missing openssl_sk_num() and openssl_sk_value()
- IdOpenSSLLoaderwasn't loading IdOpenSSLHeaders_pkcs12

mezen committed:
Added new OpenSSL IO Handler for OpenSSL 1.1.1

Added new OpenSSL 1.1.1 header translation

Added new OpenSSL IO Handler

Fixed memory corruption

If GetPAnsiChar own created UTF8String will be freed when the method
will be leaved. In that case the string is no longer accessable for
OpenSSL.

No clue what to do with the USE_MARSHALLED_PTRS part, that has the same
problem...

Implemented more compiler macros in x509

Set default values for options

Added virtual methods for context initialization for #224

Fixed small translation error

Fixed small errors found with FixInsight

Added missing empty implementations

Added Header to source files

Removed own definition of size_t and time_t

Added PPIdC_INT definition to fpc

Updated OpenSSL Header translation and Source Generator

Fixed quirk with circular references

Added IdOpenSSLConsts for better x64 support

Improved xml doc

Added missing include

Added some pem functions for read/write

Added translation for objects.h and fixed some other translations

Added missing includes of IdCompilerDefines.inc

Some FPC compability

Added  PPPByte to FPC

Implemented some compiler macros in crypto

Some more OpenSSL translation

Moved TIdOpenSSLPersistent to own unit

Added new TIdC_TM

Added new x509 wrapper and callback for verification

Extracted TIdOpenSSLVersion to own unit

Fixed problem with explicite TLS

Renamed tls version properties

Implemented better way for dynamic loading

Implemented usage of dynamic loading

mezen committed:
Added results of an internal review

mezen committed:
FTP data channel should reuse the session of the command channel

This prevents connection stealing.
See https://forum.filezilla-project.org/viewtopic.php?p=137191#p137191
for more details

mezen committed:
Added git attributes

mezen committed:
Made intermediate code more compatible

* replaced usage of reserved words, for example &in -> in_
* removed regions
* replaced accidentally added dotted unit name in crypto

Thanks to @JedrzejczykRobert

mezen committed:
Made new OpenSSL IO Handler more compatible

* removed nested consts
* replaced some not existing functions
* removed regions
* replaced $Raise with Raise_

Thanks to @JedrzejczykRobert

mezen committed:
Removed some unnecessary stuff from GenerateCode.dproj

Automated by ProjectMagican :)

mezen committed:
Do not use SafeLoadLibrary without second parameter

See for more information:
https://blog.dummzeuch.de/2017/06/13/do-not-call-safeloadlibrarydllname-dll/

mezen committed:
Added code for loading OpenSSL on Linux

Thanks to @xjikka

mezen committed:
Replaced %LINE% with %LINENUM%

%LINE% is a string, but we need an integer

Thanks to @grahamegrieve

mezen committed:
IndyCheckWindowsVersion is only available under windows

Obviously...

Thanks to @grahamegrieve

mezen committed:
Removed duplicated types

rsa_st & dsa_st & dh_st & ec_key_st are already in IdOpenSSLHeaders_ossl_typ

Thanks to @grahamegrieve

mezen committed:
If the options of the IO handler change, the context must be reinitialized

mezen committed:
Corrected some header imports

mezen committed:
Used correct case

mezen committed:
Fixed another wrong type

mezen committed:
Fixed copy'n'paste error :(

mezen committed:
Added missing raise

mezen committed:
Fixed small errors in crypto

* Missing cdecl
* Using PIdAnsiChar instead of PAnsiChar

mezen committed:
Added CRYPTO_set_mem_functions to crypto

mezen committed:
If OpenSSL is dynamically loaded, the DLLs should also be freed at unloading

mezen committed:
Fixed MemLeak with handling TLS Sessions

With new_session_cb we tell OpenSSL that we are holding a reference to
the session. We must then also release this reference. But since we get
more than one session object we need to keep a list of them.
And when cloning an IO Handler, the session reference must be increased.

Thanks to @ralfjunker

mezen committed:
Fixed line endings

mezen committed:
Fixed missing _ to parameter named "out" & added dynlibs to Loader

Thanks to @JedrzejczykRobert

mezen committed:
Fixed used-after-free

The ssl object should be destroyed before its ssl context object.

While destroying ssl context, OpenSSL calls remove_session_cb, but our
FSessionList is already freed

Thanks to @ralfjunker

mezen committed:
The Server IO Handler does not immediately start a TLS handshake any longer

You need to set PassThrough to False

Thanks to @rlebeau

mezen committed:
Added SaveToFile and ctor overload for file loading to TIdOpenSSLX509

mezen committed:
Added missing file header to IdOpenSSLX509

mezen committed:
Fixed using PBIO instead of BIO directly

mezen committed:
Use existing CheckForError and RaiseError

mezen committed:
Add additional Infos to EIdOpenSSLAcceptError

mezen committed:
Replaced `&parameter ` with `parameter_`

For better support of older delphi versions and fpc

mezen committed:
Improved excpetion message for shutdown errors

mezen committed:
Fixed multiple `Load` / `Unload` calls

mezen committed:
Resolved c macro for ASN1_INTEGER functions

mezen committed:
Fixed wrong parameter definitin in bn

mezen committed:
Resolved c macro for ASN1_Time functions

mezen committed:
Resolved c macro for ASN1_UTCTIME & ASN1_GENERALIZEDTIME functions

mezen committed:
Resolved c macro for BIO_get/set_mem functions

mezen committed:
Removed duplicate type X509_REVOKED

mezen committed:
Fixed small memory leak

Commit:
Fix compile errors for Delphi 7

Signed-off-by: Fabian S. Biehn <fbiehn@aagon.com>

mezen committed:
Added initialization of result as addition to 9c37992275a514eb1c80369a78e574d745f0646c

mezen committed:
Added BeforeInitContext and AfterInitContext to TIdOpenSSLIOHandlerServer

mezen committed:
Added some xml doc

mezen committed:
Removed old code

mezen committed:
Use parameter instead of property

mezen committed:
TLS Client is now able to verify host name

mezen committed:
Added small helpful comment

mezen committed:
Added header translation for pkcs12.h

mezen committed:
Sort uses

mezen committed:
Use newer function

mezen committed:
Added possibility to load trusted certs

Trusted certs are a special OpenSSL format

mezen committed:
Load x509 from memory directly

mezen committed:
Fixed line endings and rebase error

mezen committed:
Fixed translation error in bio

Used BIO_METHOD (wrong) instead of PBIO_METHOD (correct)

mezen committed:
Added missing uses

mezen committed:
Expanded some c compiler macro

mezen commented:
Could you please rebase your changes and add a proper description what you changed?

Delphi-FPC-Lazarus committed:
fixed warning cause of comparison of unsigned and signed integer

Delphi-FPC-Lazarus committed:
* IdOpenSSLIOHandlerClient
function GetClientSocket: TIdOpenSSLSocketClient; // {$IFDEF USE_INLINE}inline;{$ENDIF} // inline leads to a signature change between debug/release

This will lead to an F2051 on complex nested package structures containing runtime and design packages....
For now, I removed {$IFDEF USE_INLINE}inline;{$ENDIF} and everething is working, but there should be an better solution.

Delphi-FPC-Lazarus committed:
* Added properties for accessing context.
Needed for calling SSL_get_*() functions like SSL_get_version()

example:
SSL_get_version(TIdOpenSSLIOHandlerClient(IdHTTP1.IOHandler).SSLSocked.SSL

Another solution would make TIdOpenSSLContextClient.create() -> TIdOpenSSLSocket.Create() for flexible
and give an possibility to introduce own OpenSSLSocked....

Delphi-FPC-Lazarus committed:
* added missing openssl_sk_num() and openssl_sk_value()

added missing type defs for STACK, STACK_OF_GENERAL_NAME, GENERAL_NAME_union_

using example: iterate subject alt names

    cert := SSL_get_peer_certificate(TIdOpenSSLIOHandlerClient(IdHTTP1.IOHandler).SSLSocked.SSL);
    subj_name := X509_get_ext_d2i(cert, NID_subject_alt_name, nil, nil);
    for i := 0 to openssl_sk_num(PSTACK(subj_name)) - 1 do
    begin
      curname := PGENERAL_NAME(openssl_sk_value(PSTACK(subj_name), i));
      if curname.type_ = GEN_DNS then
      begin
        // if ASN1_STRING_length(PASN1_STRING(curname.d.DnsName)) = length(dns_name)
        dns_name := PAnsiChar(ASN1_STRING_get0_data(PASN1_STRING(curname.d.DnsName)));
        Memo1.Lines.Add(string(dns_name)); // Debug
      end;
    end;

- added to IntermediateCode, run generator for static and dynamic. test ok

Delphi-FPC-Lazarus committed:
fixed idopenssl loader wasn't loading IdOpenSSLHeaders_pkcs12

Delphi-FPC-Lazarus committed:
reverted duplicate declaration, sorry for confusion

Delphi-FPC-Lazarus commented:
rebase done, no conflicts any more. added description summary and detail for every change

mezen commented:
d4f2d40075d62f6052bc1df54a64fe3424f7a356 changes something out of the scope of my PR. Could you please make a PR directly to Indy for that change?

Delphi-FPC-Lazarus commented:
yea, https://github.com/mezen/Indy/commit/d4f2d40075d62f6052bc1df54a64fe3424f7a356 just fixes an ugly warning

Delphi-FPC-Lazarus committed:
Enabled IDOpenSSLhandler loading OpenSSL 3.0 dlls
Added changed SSL_get_peer_certificate()
